modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,336 @@
+	}
+
+	return &nativeHistogram{
+		metricName:       name,
+		series:           make(map[uint64]*nativeHistogramSeries),
+		onAddSerie:       onAddSeries,
+		onRemoveSerie:    onRemoveSeries,
+		traceIDLabelName: traceIDLabelName,
+		buckets:          buckets,
+	}
+}
+
+func (h *nativeHistogram) ObserveWithExemplar(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) {
+	hash := labelValueCombo.getHash()
+
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	s, ok := h.series[hash]
+	if ok {
+		h.updateSeries(s, value, traceID, multiplier)
+		return
+	}
+
+	if !h.onAddSerie(h.activeSeriesPerHistogramSerie()) {
+		return
+	}
+
+	newSeries := h.newSeries(labelValueCombo, value, traceID, multiplier)
+	h.series[hash] = newSeries
+}
+
+func (h *nativeHistogram) newSeries(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) *nativeHistogramSeries {
+	newSeries := &nativeHistogramSeries{
+		// TODO move these labels in HistogramOpts.ConstLabels?
+		labels: labelValueCombo.getLabelPair(),
+		promHistogram: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    h.name(),
+			Help:    "Native histogram for metric " + h.name(),
+			Buckets: h.buckets,
+			// TODO check if these values are sensible and break them out
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 15 * time.Minute,
+		}),
+		lastUpdated: atomic.NewInt64(0),
+	}
+
+	h.updateSeries(newSeries, value, traceID, multiplier)
+
+	return newSeries
+}
+
+func (h *nativeHistogram) updateSeries(s *nativeHistogramSeries, value float64, traceID string, multiplier float64) {
+	for i := 0.0; i < multiplier; i++ {
+		s.promHistogram.(prometheus.ExemplarObserver).ObserveWithExemplar(
+			value,
+			map[string]string{h.traceIDLabelName: traceID},
+		)
+	}
+	s.lastUpdated.Store(time.Now().UnixMilli())
+}
+
+func (h *nativeHistogram) name() string {
+	return h.metricName
+}
+
+func (h *nativeHistogram) collectMetrics(appender storage.Appender, timeMs int64, externalLabels map[string]string) (activeSeries int, err error) {
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	labelsCount := 0
+	if h.series[0] != nil {
+		labelsCount = len(h.series[0].labels.names)
+	}
+	lbls := make(labels.Labels, 1+len(externalLabels)+labelsCount)
+	lb := labels.NewBuilder(lbls)
+	activeSeries = 0
+
+	lb.Set(labels.MetricName, h.metricName)
+
+	// set external labels
+	for name, value := range externalLabels {
+		lb.Set(name, value)
+	}
+
+	for _, s := range h.series {
+
+		// Set series-specific labels
+		for i, name := range s.labels.names {
+			lb.Set(name, s.labels.values[i])
+		}
+
+		// Extract histogram
+		encodedMetric := &dto.Metric{}
+
+		// Encode to protobuf representation
+		err = s.promHistogram.Write(encodedMetric)
+		if err != nil {
+			return activeSeries, err
+		}
+		s.histogram = encodedMetric.GetHistogram()
+
+		// *** Classic histogram
```

Comment:
it looks like we are adding both a classic and native histogram to the appender? any particular reason?

Reply:
Correct.  If we don't then there isn't an opportunity to slow transition.  Without the classic histogram here, we would not be generating them at all, which would break a bunch of things.  The idea here is that we keep writing both, but likely delete the older implementation of classic histograms at a future date.

Reply:
we should talk about how a transition would work. i don't like writing a classic histo here b/c it's a bit misleading in code and the costs of doubling all histograms would be quite a lot, but i see we might need some pattern to help people migrate from one to another. 

Maybe this override is tristate instead of bool? classic/native/both?
```
MetricsGeneratorGenerateNativeHistograms
```

Reply:
I like the tristate idea.  I agree it makes it a little messy in code to have them somewhat lumped together.  At least if we had the tristate the classic histograms would be in its own block/function and clearly separated.

Reply:
I also like the tristate idea.  Currently whether you create classic histograms or not is controlled by setting the buckets on each processor, if you set them to `nil` we won't create the classic histograms. Having a setting to enable/disable it for the entire tenant in the registry would be nice :+1:

Reply:
I've pushed a commit for this.